### PR TITLE
[feature](nereids)support subquey in join condition

### PR DESCRIPTION
**Pull request by starocean999**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 1]
```diff
@@ -164,11 +169,84 @@ public List<Rule> buildRules() {
...
+                            .collect(Collectors.groupingBy(conjunct -> conjunct.containsType(SubqueryExpr.class),
+                                    Collectors.toList()));
+                    List<Expression> subqueryConjuncts = joinConjuncts.get(true);
+                    if (subqueryConjuncts == null || subqueryConjuncts.stream()
+                            .anyMatch(expr -> !isValidSubqueryConjunct(expr, join.left()))) {
+                        return join;
+                    }
+
+                    ImmutableList<Set<SubqueryExpr>> subqueryExprsList = subqueryConjuncts.stream()
+                            .map(e -> (Set<SubqueryExpr>) e.collect(SubqueryExpr.class::isInstance))
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:188]:
```suggestion
                            .Set<SubqueryExpr>map(e -> e.collect(SubqueryExpr.class::isInstance))
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 2]
```diff
@@ -164,11 +169,84 @@ public List<Rule> buildRules() {
...
+                            .collect(ImmutableList.toImmutableList());
+                    List<Expression> newConjuncts = Lists.newArrayList();
+                    LogicalPlan applyPlan = null;
+                    LogicalPlan tmpPlan = (LogicalPlan) join.left();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:192]: tmpPlan is not a good name, use a more meaningful name

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 3]
```diff
@@ -164,11 +169,84 @@ public List<Rule> buildRules() {
...
+                    List<Expression> subqueryConjuncts = joinConjuncts.get(true);
+                    if (subqueryConjuncts == null || subqueryConjuncts.stream()
+                            .anyMatch(expr -> !isValidSubqueryConjunct(expr, join.left()))) {
+                        return join;
+                    }
+
+                    ImmutableList<Set<SubqueryExpr>> subqueryExprsList = subqueryConjuncts.stream()
+                            .map(e -> (Set<SubqueryExpr>) e.collect(SubqueryExpr.class::isInstance))
+                            .collect(ImmutableList.toImmutableList());
+                    List<Expression> newConjuncts = Lists.newArrayList();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:190]: use ImmutableList.Builder<Expression>

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java [hunk 4]
```diff
@@ -164,11 +169,84 @@ public List<Rule> buildRules() {
...
+                    Plan newJoin = join.withConjunctsChildren(join.getHashJoinConjuncts(),
+                            newConjuncts, applyPlan, join.right());
+                    return newJoin;
+                }))
         );
     }
 
+    private static boolean isValidSubqueryConjunct(Expression expression, Plan leftChild) {
+        // the subquery must be uncorrelated subquery or only correlated to the left child
+        // currently only support the following 3 simple scenarios
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubqueryToApply.java:228]:
```suggestion
        // currently only support the following 4 simple scenarios
```

**Comment by starocean999** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by englefly — APPROVED** [review]

**Pull request merged**